workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/flows/RedeemToken.kt

```diff
@@ -1,85 +1,172 @@
-//        }
-//    }
-//
-//    @InitiatedBy(Initiator::class)
-//    class Responder(val otherSession: FlowSession) : FlowLogic<SignedTransaction>() {
-//        @Suspendable
-//        override fun call(): SignedTransaction {
-//            // Receive a move notification from the issuer. It tells us if we need to sign up for token updates or
-//            // generate a confidential identity.
-//            val issuanceNotification = otherSession.receive<TokenMoveNotification>().unwrap { it }
-//
-//            // Generate and send over a new confidential identity, if necessary.
-//            if (issuanceNotification.anonymous) {
-//                subFlow(RequestConfidentialIdentity.Responder(otherSession))
-//            }
-//
-//            // Resolve the issuance transaction.
-//            return subFlow(ReceiveFinalityFlow(otherSideSession = otherSession, statesToRecord = StatesToRecord.ONLY_RELEVANT))
-//        }
-//    }
-//
-//}
\ No newline at end of file
+package com.r3.corda.sdk.token.workflow.flows
+
+import co.paralleluniverse.fibers.Suspendable
+import com.r3.corda.sdk.token.contracts.states.AbstractToken
+import com.r3.corda.sdk.token.contracts.states.FungibleToken
+import com.r3.corda.sdk.token.contracts.states.NonFungibleToken
+import com.r3.corda.sdk.token.contracts.types.TokenType
+import com.r3.corda.sdk.token.workflow.selection.TokenSelection
+import com.r3.corda.sdk.token.workflow.selection.generateExitNonFungible
+import com.r3.corda.sdk.token.workflow.utilities.ownedTokenAmountWithIssuerCriteria
+import com.r3.corda.sdk.token.workflow.utilities.ownedTokensByTokenIssuer
+import net.corda.core.contracts.Amount
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.flows.CollectSignaturesFlow
+import net.corda.core.flows.FinalityFlow
+import net.corda.core.flows.FlowLogic
+import net.corda.core.flows.FlowSession
+import net.corda.core.flows.InitiatedBy
+import net.corda.core.flows.InitiatingFlow
+import net.corda.core.flows.ReceiveFinalityFlow
+import net.corda.core.flows.ReceiveStateAndRefFlow
+import net.corda.core.flows.SendStateAndRefFlow
+import net.corda.core.flows.SignTransactionFlow
+import net.corda.core.flows.StartableByRPC
+import net.corda.core.identity.Party
+import net.corda.core.node.StatesToRecord
+import net.corda.core.node.services.IdentityService
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import net.corda.core.utilities.ProgressTracker
+import net.corda.core.utilities.unwrap
+
+object RedeemToken {
+
+    @CordaSerializable
+    data class TokenRedeemNotification<T : TokenType>(val anonymous: Boolean, val amount: Amount<T>?)
+
+    // TODO later split it into non fungible or fungible
+    // Called on owner side.
+    @InitiatingFlow
+    @StartableByRPC
+    class InitiateRedeem<T : TokenType>(
```

Comment:
would be nice to think about how to make these flows more modular. E.g. they mutate tx builders.